src/cuda_scalar.cu

```diff
@@ -0,0 +1,986 @@
+
+extern "C"
+void cuda_scalar_malloc_host(void)
+{
+  checkCudaErrors(cudaMallocHost(&s, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMallocHost(&s0, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMallocHost(&s_conv, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMallocHost(&s_conv0, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMallocHost(&s_diff, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMallocHost(&s_diff0, dom[rank].Gcc.s3b * sizeof(real)));
+    cpumem += dom[rank].Gcc.s3b * sizeof(real);
+}
+
+extern "C"
+void cuda_scalar_malloc_dev(void)
+{
+  checkCudaErrors(cudaMalloc(&_bc_s, sizeof(BC_s)));
+    gpumem += sizeof(BC_s);
+  checkCudaErrors(cudaMemcpy(_bc_s, &bc_s, sizeof(BC_s), 
+    cudaMemcpyHostToDevice));
+
+  checkCudaErrors(cudaMalloc(&_s, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMalloc(&_s0, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMalloc(&_s_conv, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMalloc(&_s_conv0, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMalloc(&_s_diff, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+  checkCudaErrors(cudaMalloc(&_s_diff0, dom[rank].Gcc.s3b * sizeof(real)));
+    gpumem += dom[rank].Gcc.s3b * sizeof(real);
+
+  checkCudaErrors(cudaMemset(_s, 0., dom[rank].Gcc.s3b * sizeof(real)));
+  checkCudaErrors(cudaMemset(_s0, 0., dom[rank].Gcc.s3b * sizeof(real)));
+  checkCudaErrors(cudaMemset(_s_conv, 0., dom[rank].Gcc.s3b * sizeof(real)));
+  checkCudaErrors(cudaMemset(_s_conv0, 0., dom[rank].Gcc.s3b * sizeof(real)));
+  checkCudaErrors(cudaMemset(_s_diff, 0., dom[rank].Gcc.s3b * sizeof(real)));
+  checkCudaErrors(cudaMemset(_s_diff0, 0., dom[rank].Gcc.s3b * sizeof(real)));
+}
+
+extern "C"
+void cuda_scalar_push(void)
+{
+  checkCudaErrors(cudaMemcpy(_s, s, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+  checkCudaErrors(cudaMemcpy(_s0, s0, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+  checkCudaErrors(cudaMemcpy(_s_conv, s_conv, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+  checkCudaErrors(cudaMemcpy(_s_conv0, s_conv0, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+  checkCudaErrors(cudaMemcpy(_s_diff, s_diff, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+  checkCudaErrors(cudaMemcpy(_s_diff0, s_diff0, dom[rank].Gcc.s3b * sizeof(real),
+    cudaMemcpyHostToDevice));
+
+  int s_nn[64] = {0,
```

Comment:
Similar questions as before - any reason why hardcode to 64?

Reply:
see comment above